### PR TITLE
edited limit to 5, edited days, added extended stats for export

### PR DESCRIPTION
**Pull request by MarvelGa**

### CODE REVIEW

File: src/main/java/com/exadel/discountwebapp/statistics/controller/StatisticsController.java [hunk 1]
```diff
@@ -27,21 +27,21 @@
     @GetMapping
     public SummaryStatisticsDTO getStatistics(@RequestParam(value = "dateFrom") @DateTimeFormat(pattern = "dd.MM.yyyy", iso = DateTimeFormat.ISO.DATE) LocalDate dateFrom,
                                               @RequestParam(value = "dateTo") @DateTimeFormat(pattern = "dd.MM.yyyy", iso = DateTimeFormat.ISO.DATE) LocalDate dateTo) {
-        return statisticsService.getStats(dateFrom.atStartOfDay(), dateTo.atStartOfDay());
+        return statisticsService.getStats(dateFrom.atStartOfDay(), dateTo.atStartOfDay().plusHours(24));
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/statistics/controller/StatisticsController.java]: why do you add 24 hours?

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java [hunk 2]
```diff
@@ -39,6 +45,68 @@ public SummaryStatisticsDTO getStats(LocalDateTime dateFrom, LocalDateTime dateT
...
+    public List<ExtendedUserVO> getExtendedDataDiscountPerUsers(LocalDateTime dateFrom, LocalDateTime dateTo) {
+        var extendedUserData = userDiscountRepository.getExtendedUserDiscountStatistics(dateFrom, dateTo);
+
+        List<ExtendedUserVO> result = new ArrayList<>();
+        for (ExtendedUserDTO elem : extendedUserData) {
+            result.add(statsMapper.userToVO(elem));
+        }
+        return result.stream()
+                .sorted((v1, v2) -> (int) (v2.getQuantity() - v1.getQuantity()))
+                .collect(Collectors.toList());
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: It is possible to do all of this with a single stream
extendedUserData.stream()
.map(statsMapper::userToVO)
.sorted(...)
.collect(...)

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java [hunk 3]
```diff
@@ -39,6 +45,68 @@ public SummaryStatisticsDTO getStats(LocalDateTime dateFrom, LocalDateTime dateT
...
+    public List<ExtendedCategoryVO> getExtendedDataActivatedDiscountPerCategory(LocalDateTime dateFrom, LocalDateTime dateTo) {
+        var categoryData = userDiscountRepository.getExtendedCategoryDiscountStatistics(dateFrom, dateTo);
+
+        List<ExtendedCategoryVO> result = new ArrayList<>();
+        for (ExtendedCategoryDTO elem : categoryData) {
+            result.add(statsMapper.categoryToVO(elem));
+        }
+        return result.stream()
+                .sorted((v1, v2) -> (int) (v2.getQuantity() - v1.getQuantity()))
+                .collect(Collectors.toList());
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: same here

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java [hunk 4]
```diff
@@ -39,6 +45,68 @@ public SummaryStatisticsDTO getStats(LocalDateTime dateFrom, LocalDateTime dateT
...
+        var vendorData = userDiscountRepository.getExtendedVendorDiscountStatistics(dateFrom, dateTo);
+
+        List<ExtendedVendorVO> result = new ArrayList<>();
+        for (ExtendedVendorDTO elem : vendorData) {
+            result.add(statsMapper.vendorToVO(elem));
+        }
+
+        return result.stream()
+                .sorted((v1, v2) -> (int) (v2.getQuantity() - v1.getQuantity()))
+                .collect(Collectors.toList());
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: same here

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java [hunk 5]
```diff
@@ -39,6 +45,68 @@ public SummaryStatisticsDTO getStats(LocalDateTime dateFrom, LocalDateTime dateT
...
+        List<ExtendedDiscountVO> result = new ArrayList<>();
+
+        for (ExtendedDiscountViewsDTO elem : disViewsData) {
+            if (elem.getViewNumber() != null)
+                result.add(statsMapper.discountToVO(elem));
+        }
+
+        return result.stream()
+                .sorted((v1, v2) -> (int) (v2.getViewNumber() - v1.getViewNumber()))
+                .collect(Collectors.toList());
```
Comment (dmitriychernov7) [inline src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: same here

File: src/main/java/com/exadel/discountwebapp/discount/repository/DiscountRepository.java [hunk 6]
```diff
@@ -16,4 +17,8 @@
     @Query(value = "SELECT new com.exadel.discountwebapp.statistics.dto.DiscountViewingDTO(d.id, d.title, d.viewNumber)" +
             " FROM Discount d")
     List<DiscountViewingDTO> getDiscountSummary();
+
+    @Query(value = "SELECT new com.exadel.discountwebapp.statistics.dto.extendeddto.ExtendedDiscountViewsDTO(d.title, d.shortDescription, d.description, d.promocode, d.percentage, d.flatAmount, d.startDate, d.expirationDate, d.vendor.title, d.category.title, d.viewNumber)" +
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/repository/DiscountRepository.java]: let's remove custom projection

File: src/main/java/com/exadel/discountwebapp/statistics/excelexport/XLSXExported.java [hunk 7]
```diff
@@ -19,136 +15,165 @@
...
+        } else if (value instanceof Integer) {
+            cell.setCellValue((Integer) value);
         } else {
             cell.setCellValue((String) value);
         }
         cell.setCellStyle(style);
     }
 
     private void writeDataLines() {
+        writeHeaderLine("users");
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/statistics/excelexport/XLSXExported.java]: move "users" to constants

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java [hunk 8]
```diff
@@ -49,13 +117,13 @@ public SummaryStatisticsDTO getStats(LocalDateTime dateFrom, LocalDateTime dateT
 
         result = result.stream()
                 .sorted((v1, v2) -> (int) (v2.getQuantity() - v1.getQuantity()))
-                .limit(10)
+                .limit(5)
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: let's move to a separate param

File: src/main/java/com/exadel/discountwebapp/statistics/controller/StatisticsController.java (hunk 1 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/statistics/controller/StatisticsController.java]: it's for set range date during one day, for example, dateFrom=16.07.2021&dateTo=16.07.2021, where dateFrom -start of the day and dateTo - end of the day

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java (hunk 2 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: fixed

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java (hunk 3 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: fixed

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java (hunk 4 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: fixed

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java (hunk 5 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: fixed

File: src/main/java/com/exadel/discountwebapp/statistics/excelexport/XLSXExported.java (hunk 7 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/statistics/excelexport/XLSXExported.java]: fixed

File: src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java (hunk 8 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/statistics/service/StatisticsService.java]: fixed

File: src/main/java/com/exadel/discountwebapp/discount/repository/DiscountRepository.java (hunk 6 above)
Comment (MarvelGa) [reply src/main/java/com/exadel/discountwebapp/discount/repository/DiscountRepository.java]: I try did it but where was issue (Failed to convert from type [java.lang.Object[]] to type [@org.springframework.data.jpa.repository.Query com.exadel.discountwebapp.discount.entity.Discount]), maybe cuz some data from this query not fit to Discount entity. maybe through additional fields such as (d.vendor.title, d.category.title). So I left the previous version

**Review by ahlukhouski2 — APPROVED** [review]

**Review by Bozhokmaria — APPROVED** [review]

**Pull request merged**